adam/visualization/positioning.py

```diff
@@ -0,0 +1,426 @@
+    def get_min_max_corner_projections(projections: torch.Tensor):
+        """
+        Retrieve the minimum and maximum corner projection (min/max extent in that dimension) for each axis
+        Args:
+            projections: Tensor(3, 8) -> corner projections onto each of three dimensions
+
+        Returns:
+            Tensor(3, 2) -> (min, max) values for each of three dimensions
+
+        """
+        check_arg(projections.shape == (3, 8))
+
+        min_indices = torch.min(projections, 1)
+        max_indices = torch.max(projections, 1)
+        # these are tuples of (values, indices), both of which are tensors
+
+        # helper variable for representing dimension numbers
+        # see https://github.com/pytorch/pytorch/issues/24807 re: pylint issue
+        dims = torch.tensor([0, 1, 2], dtype=torch.int)  # pylint: disable=not-callable
+        # select the indexed items (from a 24 element tensor)
+        minima = torch.take(projections, min_indices[1] + (dims * 8))
+        maxima = torch.take(projections, max_indices[1] + (dims * 8))
+        # stack the minim
+        return torch.stack((minima, maxima), 1)
+
+    @staticmethod
+    def get_min_max_overlaps(
+        min_max_proj_0: torch.Tensor, min_max_proj_1: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Given min/max corner projections onto 3 axes from two different objects,
+        return an interval for each dimension representing the degree of overlap or
+        separation between the two objects.
+        Args:
+            min_max_proj_0: Tensor(3,2) min_max_projections for box 0
+            min_max_proj_1: Tensor(3,2) min_max projections for box 1
+
+        Returns:
+            (3, 2) tensor -> ranges (start, end) of overlap in each of three dimensions
+        """
+        check_arg(min_max_proj_0.shape == (3, 2))
+        check_arg(min_max_proj_1.shape == (3, 2))
+
+        # see https://github.com/pytorch/pytorch/issues/24807 re: pylint issue
+        dims = torch.tensor([0, 1, 2], dtype=torch.int)  # pylint: disable=not-callable
+
+        mins_0 = min_max_proj_0.gather(
+            1, torch.tensor([[0], [0], [0]])  # pylint: disable=not-callable
+        )
+        mins_1 = min_max_proj_1.gather(
+            1, torch.tensor([[0], [0], [0]])  # pylint: disable=not-callable
+        )
+
+        combined_mins = torch.stack((mins_0, mins_1), 1).squeeze()
+        max_indices = torch.max(combined_mins, 1)
+        maximum_mins = torch.take(combined_mins, max_indices[1] + (dims * 2))
+
+        # should stick together the minimum parts and the maximum parts
+        # with columns like:
+        # [ min0x   min1x
+        #   min0y   min1y
+        #   min0z   min1z
+        #                ]
+        # then find the maximum element from each row
+
+        # repeat the process for the min of the max projections
+        maxs_0 = min_max_proj_0.gather(
+            1, torch.tensor([[1], [1], [1]])  # pylint: disable=not-callable
```

Comment:
here and below, `ones`?

Reply:
I would need the shorthand to turn a flat vector of [1, 1, 1,] into 
[
    [1],
    [1],
    [1]
]

Reply:
which would be torch.zeros(3, 1) 😅